modules/openapi-generator/src/main/resources/typescript-fetch/runtime.mustache

```diff
@@ -265,6 +299,8 @@ export interface ResponseContext {
 export interface Middleware {
     pre?(context: RequestContext): Promise<FetchParams | void>;
     post?(context: ResponseContext): Promise<Response | void>;
+    fail?(context: RequestContext, err: unknown): Promise<unknown | void>;
+    complete?(context: RequestContext): void;
```

Comment:
While I can't say I'm a big fan of the original "pre/post" middleware design, the newly introduced fail and complete do not seem to be consistent with those:
- "fail" could take ErrorContext instead of an extra positional parameter
- "complete" is synchronous

The new middlewares bring up some new challenges, too. E.g. it's not possible to implement retry logic with the "fail" middleware - because no matter what the middleware does, the end result is a rejected promise. Also one cannot really chain multiple fail middlewares, since they all receive the same original context and error (compare that to "post" middlewares that can implement chained transformations one on top of another - because the output of each middleware is used to form the input for the next one in the chain).

I feel that the suggested approach adds more complexity to manage than it solves problems.

And frankly, I would even suggest revising the existing pre/post middlewares into something more universal. Angular's http interceptors can be a good example of flexible design. Or middlewares in Express.js. In both those examples, only one type of middleware exists, but it can be used to implement both pre and post behavior as well as custom error handling etc.

Reply:
> And frankly, I would even suggest revising the existing pre/post middlewares into something more universal.

Refactoring the middlewares will bring us some breaking changes and impact those who already using this template. that's not what i expected.

> Or middlewares in Express.js. In both those examples, only one type of middleware exists, but it can be used to implement both pre and post behavior as well as custom error handling etc.

“Only one type of middleware exists” means you don't know which phase the request is currently in.  so in angular, we have to write something like this:

```
return next.handle(request).pipe(
  tap((event: HttpEvent<any>) => {
  if (event instanceof HttpResponse) {
    console.log("Your request finished");
    }
  }, (err: any) => {
    console.log("You get an error.");
  })
);
```

It looks more complicated and unintuitive.

So I may implement it into two separate options: listener and middleware.

Listener will act like its name, it will only listen events and ***CAN NOT*** interfere the request process.

Middleware can modify the request before it was sent and transform the response after it was succeeded.

> The new middlewares bring up some new challenges, too. E.g. it's not possible to implement retry logic with the "fail" middleware

As for the retry logic, I think we shouldn't handle it in this tool because it's too complicated.

When we retry a request, a completely new request was launched actually and all processes should be performed including those customer's logic in their code.

Reply:
> “Only one type of middleware exists” means you don't know which phase the request is currently in

that's not correct. a middleware has ultimate control over that: it runs its own logic either before calling the next handler in the chain or after that.

Reply:
I also disagree, a Retry policy should be handled by the library, it's only boilerplate for customers use case. If needed a way to call without a retry should be possible however then.